text/1028-cli-hooks.rst

```diff
@@ -56,70 +69,182 @@ are always executed in WSL.
+file in the ``[watch]`` table. We will introduce a ``[watch.files]`` sub-table
+for watching file system changes and triggering scripts.
+
+The general structure of watch tables is going to use the key (potentially
+quoted) to specify *what is being watched* and the corresponding value to
+specify the script that will be triggered by changes.
+
+The ``gel watch`` command (without needing further options) would then be used
+to start the watch process and monitor whatever is specified in the
+``gel.toml``. By default ``gel watch`` will only watch the files specified in
+the ``gel.toml`` config and execute the trigger scripts. Running ``gel watch
+--migrate`` will additionally monitor schema changes and perform real-time
+migrations.
+
+This is a backwards incompatible change compared to how ``edgedb watch`` operates
+now.
+
+The output of the scripts will then appear in the same TTY as the ``gel
+watch`` command.
+
+Multiple consecutive file system change events will be debounced into a single
+event. For example, given scripts ``src/**`` and ``*.py`` and writes to
+``src/a.py``, ``src/b.py`` and ``src/c.rs``, both scripts will trigger exactly
+once. To achive this, ``gel watch``` will wait for a small amount of time
+(100ms) after receiving first event before executing matching scripts. This is
+mostly to reduce unnecessary multiple triggers for the same watched entity.
+
+Another consequence of executing triggered scripts in the background is that
+sometimes the changes are invalid in some way and the script will fail. This
+is considered part of normal operation (e.g. syntax error in a saved query
+file) and the scripts should be such that failure does not create some
+non-recoverable state.
+
+Due to the nature of monitoring changes, we cannot guarantee any particular
+order in which the watch scripts will be triggered when multiple changes occur
+at once. This is independent of the source of multiple triggers: whether it is
+due to multiple files being updated or multiple watch rules matching the same
+file. The scripts are effectively triggered asynchronously and the exact order
+is an implementation detail that should not be relied upon.
 
-The ``gel branch`` commands may also trigger these scripts since different
-branches can have different schemas. The ``switch``, ``rebase``, ``merge``,
-and ``wipe`` branch commands all potentially change the current schema. This
-means that the hooks associated with applying schema changes must be executed
-for these commands as well.
+
+Files
+-----
+
+The ``[watch.files]`` table should contain keys that are interpreted as file
+system paths. They can also contain common glob patterns (such as provided by
+`this library <https://docs.rs/globset/latest/globset/#syntax>`_).
+
+The paths should be valid in the underlying file system (therefore using ``/``
+for Linux and ``\`` for Windows, etc.). The relative file paths are assumed to
```

Comment:
This is really inconvenient because it makes `gel.toml` incompatible between OSs. Can we make it always use `/`, even on Windows?

Reply:
Yeah, I can phrase it to be always `/`. The compatibility issue is important.